iphone/CoreApi/CoreApi/Logger/Logger.mm

```diff
@@ -1,32 +1,231 @@
+    }
+  }
+}
+
++ (uint64_t)getLogFileSize {
+  Logger * logger = [self logger];
+  return logger.fileHandle != nil ? [logger.fileHandle offsetInFile] : 0;
+}
+
+// MARK: - C++ injection
+
+void LogMessage(base::LogLevel level, base::SrcPoint const & src, std::string const & message)
+{
+  [Logger logMessageWithLevel:level src:src message:message];
+  CHECK_LESS(level, base::g_LogAbortLevel, ("Abort. Log level is too serious", level));
+}
+
+bool AssertMessage(base::SrcPoint const & src, std::string const & message)
+{
+  [Logger logMessageWithLevel:base::LCRITICAL src:src message:message];
+  return true;
+}
+
+// MARK: - Private
+
++ (void)enableFileLogging {
+  Logger * logger = [self logger];
+  NSFileManager * fileManager = [NSFileManager defaultManager];
+
+  // Create a log file if it doesn't exist and setup file handle for writing.
+  if (![fileManager fileExistsAtPath:kLogFilePath])
+    [fileManager createFileAtPath:kLogFilePath contents:nil attributes:nil];
+  NSFileHandle * fileHandle = [NSFileHandle fileHandleForWritingAtPath:kLogFilePath];
+  if (fileHandle == nil) {
+    LOG(LERROR, ("Failed to open log file for writing", kLogFilePath.UTF8String));
+    [self disableFileLogging];
+    return;
+  }
+  // Clean up the file if it exceeds the maximum size.
+  if ([fileManager contentsAtPath:kLogFilePath].length > kMaxLogFileSize)
+    [fileHandle truncateFileAtOffset:0];
+
+  logger.fileHandle = fileHandle;
+
+  _fileLoggingEnabled = YES;
+  LOG(LINFO, ("File logging is enabled"));
+}
+
++ (void)disableFileLogging {
+  Logger * logger = [self logger];
+
+  [logger.fileHandle closeFile];
+  logger.fileHandle = nil;
+  [self removeFileAtPath:kLogFilePath];
+
+  _fileLoggingEnabled = NO;
+  LOG(LINFO, ("File logging is disabled"));
+}
+
++ (void)logMessageWithLevel:(base::LogLevel)level src:(base::SrcPoint const &)src message:(std::string const &)message {
+  // Build the log message string.
+  auto & logHelper = base::LogHelper::Instance();
+  std::ostringstream output;
+  logHelper.WriteProlog(output, level);
+  logHelper.WriteLog(output, src, message);
+  auto const logMessage = output.str();
+
+  Logger * logger = [self logger];
+  // Log the message into the system log.
+  os_log(logger.osLogger, "%{public}s", logMessage.c_str());
+
+  dispatch_async([self fileLoggingQueue], ^{
+    // Write the log message into the file.
+    NSFileHandle * fileHandle = logger.fileHandle;
+    if (fileHandle != nil) {
+      [fileHandle seekToEndOfFile];
+      [fileHandle writeData:[NSData dataWithBytes:logMessage.c_str() length:logMessage.length()]];
+    }
+  });
+}
+
++ (NSURL *)getZippedLogFile:(NSString *)logFilePath {
+  NSString * zipFilePath = [[logFilePath stringByDeletingPathExtension] stringByAppendingPathExtension:kZipLogFileExtension];
+  auto const success = CreateZipFromFiles({logFilePath.UTF8String}, zipFilePath.UTF8String);
+  if (!success) {
+    LOG(LERROR, ("Failed to zip log file:", kLogFilePath.UTF8String));
+    return nil;
```

Comment:
Is it possible to return an original unzipped file if zipping has failed?

Reply:
Like the idea. Thank!